## Use Redit API to Aquire Study Data

### Data from redit/democrats

In the following sections, the structure of the data is explored in order to allow the construction of a for loop to read in the data and output in a csv format.

In [105]:
import requests   #importing required Dictionaries
import pandas as pd
import json
import time

In [109]:
url = 'https://www.reddit.com/r/democrats/.json'

In [80]:
res = requests.get(url, headers = {'User-agent': 'Snakes'})

In [81]:
res.status_code

200

### Sort out the Structure of the Redit data

In [82]:
data = res.json()   #get back information as a dictionary set it equal to data

In [83]:
data.keys()

dict_keys(['kind', 'data'])

In [84]:
data['kind']  #kind is not important 

'Listing'

In [85]:
data['data'].keys()  #interested in data and it is a dictionary with the following keys

dict_keys(['modhash', 'dist', 'children', 'after', 'before'])

In [86]:
data['data']['modhash'] # Seems to be an empty string



''

In [87]:
data['data']['dist']  #an integer 26 the length of the list

26

In [88]:
data['data']['children'][0]   #list of objects

{'data': {'approved_at_utc': None,
  'approved_by': None,
  'archived': False,
  'author': 'backpackwayne',
  'author_flair_background_color': None,
  'author_flair_css_class': None,
  'author_flair_richtext': [],
  'author_flair_template_id': None,
  'author_flair_text': None,
  'author_flair_text_color': None,
  'author_flair_type': 'text',
  'author_fullname': 't2_3ci9p',
  'banned_at_utc': None,
  'banned_by': None,
  'can_gild': False,
  'can_mod_post': False,
  'category': None,
  'clicked': False,
  'content_categories': None,
  'contest_mode': False,
  'created': 1533708499.0,
  'created_utc': 1533679699.0,
  'distinguished': 'moderator',
  'domain': 'self.democrats',
  'downs': 0,
  'edited': False,
  'gilded': 0,
  'hidden': False,
  'hide_score': False,
  'id': '95g5w2',
  'is_crosspostable': False,
  'is_meta': False,
  'is_original_content': False,
  'is_reddit_media_domain': False,
  'is_self': True,
  'is_video': False,
  'likes': None,
  'link_flair_background_color': '

In [89]:
data['data']['after'] #Location string

't3_9cofaf'

In [90]:
data['data']['before']  #empty right now

In [91]:
print(len(data['data']['children']))

26


In [92]:
redit_posts = [post['data'] for post in data['data']['children']]

pd.DataFrame(posts).head()

approved_at_utc approved_by  archived           author  \
0            None        None     False         tehForce   
1            None        None     False  thatrightwinger   
2            None        None     False        chabanais   
3            None        None     False       optionhome   
4            None        None     False           sylect   

  author_flair_background_color author_flair_css_class  \
0                                                 None   
1                                            Far-Right   
2                                                 Bold   
3                                         Conservative   
4                                      Neoconservative   

                               author_flair_richtext  \
0  [{'e': 'text', 't': 'Liberal Education != Purp...   
1          [{'e': 'text', 't': 'Don't Tread on Me'}]   
2         [{'e': 'text', 't': 'Stronger than derp'}]   
3               [{'e': 'text', 't': 'Conservative'}]   
4           [{'e': 'text', 't': 'Uranus Astronaut'}]   

               author_flair_template_id                 author_flair_text  \
0                                  None  Liberal Education != Purple Hair   
1  9b86186a-8b38-11e1-8f58-12313d2c1af1                 Don't Tread on Me   
2                                  None                Stronger than derp   
3  421b644c-8b38-11e1-bc5f-12313d18ad57                      Conservative   
4  2f6dbf84-8b38-11e1-bc5f-12313d18ad57                  Uranus Astronaut   

  author_flair_text_color ...  thumbnail_height thumbnail_width  \
0                    dark ...               NaN             NaN   
1                    dark ...             105.0           140.0   
2                    dark ...             140.0           140.0   
3                    dark ...             103.0           140.0   
4                    dark ...              73.0           140.0   

                                               title  ups  \
0    Michael Oakeshott - this week's sidebar tribute   43   
1  Watch Live: Judge Brett Kavanaugh's confirmati...   14   
2                      Ben Shapiro lays them away...  883   
3  Rev. Jasper Williams Jr..."children being in a...  591   
4  Huge Number Of Illegals Opting Out Of Welfare ...  434   

                                                 url  user_reports view_count  \
0  https://www.reddit.com/r/Conservative/comments...            []       None   
1        https://www.youtube.com/watch?v=h8il9dAhA00            []       None   
2                https://i.redd.it/i3xkb1sgb8k11.jpg            []       None   
3  https://apnews.com/2a382030f1964300ae27d05d351...            []       None   
4  https://www.dailywire.com/news/35386/report-il...            []       None   

   visited whitelist_status  wls  
0    False          all_ads    6  
1    False          all_ads    6  
2    False          all_ads    6  
3    False          all_ads    6  
4    False          all_ads    6  

[5 rows x 95 columns]

In [93]:
current_url = url + '?after' + current_dict['data']['after']
current_url

#need to go back to video not bringing in new data

'https://www.reddit.com/r/democrats/.json?aftert3_9cwoyu'

### Set a loop to get a larger quantity of post

The program can now get and understand a data request.  However more than 25 cases are needed for the analysis so created a loop to get more data.  Ran the loop multiple times to get more than the 1000 cases allowed normally.  Selected on after variable to make sure and get unique cases.

Running the Cells below should collect the data and save it to a csv file.  

In [188]:
import requests   #importing required Dictionaries
import pandas as pd
import json
import time

In [189]:
posts = [] #above code was to figure out structure  this loop should do it all


In [190]:
import requests 
after = None
url = 'https://www.reddit.com/r/democrats/.json'
#code below is the loop to get the data and save it to the dataframe

In [191]:
for _ in range(35):  # for loop to his reddit mulitiple times.  
    if after == None:
        current_url = url
    else: 
        current_url = url + '?after=' + current_dict['data']['after']    
    res = requests.get(current_url, headers = {'User-agent': 'Snake_2'}) 
    current_dict = res.json()
    if res.status_code != 200:
        print('res error ',  res.status_code)
        break
    else:
        print(len(posts), after)
   
   
    redit_posts = [post['data'] for post in current_dict['data']['children']]
    posts.extend(redit_posts)
    
    after = current_dict['data']['after']
    
    time.sleep(4)
    pd.DataFrame(posts).to_csv('./data/democratic.csv', index = False)
   
    

0 None
26 t3_9dollh
51 t3_9djch3
76 t3_9dcqw7
101 t3_9d166j
126 t3_9d24io
151 t3_9cxdic
176 t3_9cv8s6
201 t3_9cquei
226 t3_9cclz5
251 t3_9c8009
276 t3_9c9483
301 t3_9c1cfa
326 t3_9bypjm
351 t3_9bt508
376 t3_9be585
401 t3_9bhfvm
426 t3_9b6a6i
451 t3_9aq5rv
476 t3_9auh6x
501 t3_9ald3a
526 t3_9acw2u
551 t3_9a6lhe
576 t3_9a7835
601 t3_99y4kc
626 t3_99unb9
651 t3_99o3km
676 t3_99l25r
701 t3_99aodd
726 t3_998lf4
751 t3_98yjm0
776 t3_98n33j
801 t3_98qgud
826 t3_98nepr
851 t3_98lvjh


In [192]:
print(len(posts))
 

   

876


### Aquire Posts from Republican Redit 

The posts from the democratic reddit have been obtained and saved as a csv file.  Now republican redit posts must be acquired for the analysis

In [199]:
posts_r = []


In [200]:
url = 'https://www.reddit.com/r/Conservative/.json'

after_r = None

In [201]:
#set up the same for loop for conservative posts from this subredit
for _ in range(32):  # for loop to his reddit mulitiple times.  
    if after_r == None:
        current_url = url
    else: 
        current_url = url + '?after=' + after_r 
    #find new url   
    res = requests.get(current_url, headers = {'User-agent': 'Snake_2'}) 
    current_dict = res.json()
    #break if a problem
    if res.status_code != 200:
        print('res error ',  res.status_code)
        break
    else:
        print(len(posts_r), after_r)
 
    #add new posts to list  
    redit_posts = [reddit['data'] for reddit in current_dict['data']['children']]
    posts_r.extend(redit_posts)
        
    after_r = current_dict['data']['after']
    time.sleep(4) #time between scrapes
    pd.DataFrame(posts_r).to_csv('./data/republican.csv', index = False)


0 None
27 t3_9djca7
52 t3_9dlw1k
77 t3_9dnjbl
102 t3_9d8a5n
127 t3_9de5ov
152 t3_9d8mx8
177 t3_9d9fft
202 t3_9d95w1
227 t3_9cwoq1
252 t3_9cyvm5
277 t3_9d8ajw
302 t3_9d2a3g
327 t3_9csodo
352 t3_9d0ap6
377 t3_9cppz4
402 t3_9coleh
427 t3_9cm7b4
452 t3_9cldr3
477 t3_9cd8ru
502 t3_9cdhn8
527 t3_9c7kyx
552 t3_9c4c0m
577 t3_9c4mwr
602 t3_9bx9z6
627 t3_9bz6k0
652 t3_9bwm68
677 t3_9byq4q
702 t3_9bpc9j
727 t3_9bjkvq
752 t3_9bl8xa
777 t3_9bmb7z


In [202]:
len(posts_r)

799

Have 801 conservative posts saved to republican.csv.  r/Conservative circled back to the start earlier  than r/democrat but seemed to cut off earlier. There are 876 r/democrat posts. 

In [ ]:
## YOUR CODE HERE

In [ ]:
## YOUR CODE HERE